Add support for Python 3.12

danyeaw opened the pull request:
Python 3.12 is now in beta and this PR adds support.

danyeaw committed:
setup: Unblock installation on Python 3.12.

danyeaw committed:
Update opcodes for Python 3.12

danyeaw committed:
Get version with importlib metadata

bwoodsend commented:
I've got a half baked branch for the imp->importlib transition.

danyeaw commented:
@bwoodsend Oh nice! Could we get those changes merged in first, or could I cherry-pick them in to this PR?

bwoodsend commented:
Working on it...

There's also this lovely surprise [from the changelog](https://docs.python.org/3.12/whatsnew/changelog.html):

> [bpo-44113](https://bugs.python.org/issue?@action=redirect&bpo=44113): Deprecate the following functions to configure the Python initialization:
> 
> * [PySys_AddWarnOptionUnicode()](https://docs.python.org/3.12/c-api/sys.html#c.PySys_AddWarnOptionUnicode)
> * [PySys_AddWarnOption()](https://docs.python.org/3.12/c-api/sys.html#c.PySys_AddWarnOption)
> * [PySys_AddXOption()](https://docs.python.org/3.12/c-api/sys.html#c.PySys_AddXOption)
> * PySys_HasWarnOptions()
> * [Py_SetPath()](https://docs.python.org/3.12/c-api/init.html#c.Py_SetPath)
> * [Py_SetProgramName()](https://docs.python.org/3.12/c-api/init.html#c.Py_SetProgramName)
> * [Py_SetPythonHome()](https://docs.python.org/3.12/c-api/init.html#c.Py_SetPythonHome)
> * [Py_SetStandardStreamEncoding()](https://docs.python.org/3.12/c-api/init.html#c.Py_SetStandardStreamEncoding)
> * _Py_SetProgramFullPath()
>
> Use the new [PyConfig](https://docs.python.org/3.12/c-api/init_config.html#c.PyConfig) API of the [Python Initialization Configuration](https://docs.python.org/3.12/c-api/init_config.html#init-config) instead ([PEP 587](https://peps.python.org/pep-0587/)).
> 

although it looks like we're safe for another year.

rokm commented:
> Working on it...
> 
> There's also this lovely surprise [from the changelog](https://docs.python.org/3.12/whatsnew/changelog.html):
> 
> > [bpo-44113](https://bugs.python.org/issue?@action=redirect&bpo=44113): Deprecate the following functions to configure the Python initialization:
> > 
> > * [PySys_AddWarnOptionUnicode()](https://docs.python.org/3.12/c-api/sys.html#c.PySys_AddWarnOptionUnicode)
> > * [PySys_AddWarnOption()](https://docs.python.org/3.12/c-api/sys.html#c.PySys_AddWarnOption)
> > * [PySys_AddXOption()](https://docs.python.org/3.12/c-api/sys.html#c.PySys_AddXOption)
> > * PySys_HasWarnOptions()
> > * [Py_SetPath()](https://docs.python.org/3.12/c-api/init.html#c.Py_SetPath)
> > * [Py_SetProgramName()](https://docs.python.org/3.12/c-api/init.html#c.Py_SetProgramName)
> > * [Py_SetPythonHome()](https://docs.python.org/3.12/c-api/init.html#c.Py_SetPythonHome)
> > * [Py_SetStandardStreamEncoding()](https://docs.python.org/3.12/c-api/init.html#c.Py_SetStandardStreamEncoding)
> > * _Py_SetProgramFullPath()
> > 
> > Use the new [PyConfig](https://docs.python.org/3.12/c-api/init_config.html#c.PyConfig) API of the [Python Initialization Configuration](https://docs.python.org/3.12/c-api/init_config.html#init-config) instead ([PEP 587](https://peps.python.org/pep-0587/)).
> 
> although it looks like we're safe for another year.

That doesn't look too bad. Last time I looked at the interpreter initialization/configuration, I was (mistakenly?) under impression we would have to deal with config struct directly, which would be a pain to do in version agnostic way. But this looks like an opaque structure with an API that abstracts the actual struct layout away, so it's quite nice for our use case.

It's available from python 3.8 on, so we can start making the switch once we drop support for 3.7.

rokm committed:
depend: bytecode: fix oparg handling for LOAD_ATTR under python 3.12

Python 3.12 removed LOAD_METHOD opcode; LOAD_ATTR now behaves like
LOAD_METHOD if the low bit of its oparg is set. This means that
LOAD_ATTR pushes namei>>1 on stack. (This change is similar to
what python 3.11 did with LOAD_GLOBAL).

bwoodsend commented:
#7676 is merged, let's see where a rebase takes us...

bwoodsend commented:
... but not using a merge commit!!!

rokm commented:
It will eventually take us here: https://github.com/rokm/pyinstaller/actions/runs/5166038863

Overall, it's not that terrible...

---

``` 
FAILED tests/unit/test_bytecode.py::test_nested_codes - AssertionError: assert [('bar', []),...('fizz', [3])] == [('bar', []), ('range', [10])]
  Left contains one more item: ('fizz', [3])
  Full diff:
  - [('bar', []), ('range', [10])]
  + [('bar', []), ('range', [10]), ('fizz', [3])]
  ?   
```

I'm not sure what this one is about yet...

---

``` 
FAILED tests/functional/test_basic.py::test_distutils_submod[onedir] - Failed: Running exe /Users/runner/work/_temp/pytest-of-runner/pytest-0/popen-gw0/test_distutils_submod_onedir_0/dist/test_source/test_source failed with return-code 1.
FAILED tests/functional/test_basic.py::test_distutils_submod[onefile] - Failed: Running exe /Users/runner/work/_temp/pytest-of-runner/pytest-0/popen-gw0/test_distutils_submod_onefile_0/dist/test_source failed with return-code 1.
FAILED tests/functional/test_basic.py::test_python_makefile[onedir-from distutils ] - Failed: Running exe /Users/runner/work/_temp/pytest-of-runner/pytest-0/popen-gw0/test_python_makefile_onedir_fr0/dist/test_source/test_source failed with return-code 1.
FAILED tests/functional/test_basic.py::test_python_makefile[onefile-from distutils ] - Failed: Running exe /Users/runner/work/_temp/pytest-of-runner/pytest-0/popen-gw0/test_python_makefile_onefile_f0/dist/test_source failed with return-code 1.
``` 

These tests using `distutils` should *not* be failing as `distutils` from `setuptools` should be available. 

My current hypothesis is that the current modulegraph's approach of looping over search locations, taking importer for it, and trying to look up the module for it was working for <= 3.11, because stdlib `distutils` was discoverable by built-in `FileFinder` (for example, in /usr/lib64/python3.11/distutils/__init__.py`, so the stdlib module was found, and then the hook took care of collecting setuptools' variant.

With 3.12, we don't have stdlib `distutils` anymore, so location-found finder's fail to find the module, and the hook is not ran at all.

I'll see if we can completely replace this block:

https://github.com/pyinstaller/pyinstaller/blob/fd0a25a718a18b0cac803cb0ec6cc5950bc8229c/PyInstaller/lib/modulegraph/modulegraph.py#L2952-L3026

with `importlib.util.find_spec`, which should work for all meta-finders and ignore the search paths...

---

``` 
FAILED tests/functional/test_import.py::test_import_metapath1[onedir] - Failed: Running exe /tmp/pytest-of-runner/pytest-0/popen-gw1/test_import_metapath1_onedir_0/dist/test_source/test_source failed with return-code 1.
FAILED tests/functional/test_import.py::test_import_metapath1[onefile] - Failed: Running exe /tmp/pytest-of-runner/pytest-0/popen-gw1/test_import_metapath1_onefile_0/dist/test_source failed with return-code 1.
``` 

I think these are caused by lack of `find_spec` method on the custom `VendoredImporter` that the test use. The custom importer seems to be taken from `setuptools`, so we can try syncing with its latest version.

rokm commented on .github/workflows/ci.yml:
```diff
@@ -116,7 +116,7 @@ jobs:
 
       - name: Install test dependencies (libraries)
-        if: ${{ !endsWith(matrix.python-version, '-dev') }}
+        if: matrix.python-version != '3.12'
```
The "proper" way would be to use `python-version: [..., '3.12-dev']` until the release and until we can install full environment.

rokm committed:
setup: python 3.12 requires pywin32-ctypes from git main

The latest release of pywin32-ctypes (0.2.0) is incompatible with
python 3.12, so we need to install the version from git main
branch. This version updated the custom module path finder to
be compliant with PEP 451.

rokm committed:
tests: fix test_import_metapath1 for python 3.12

Update the `VendorImporter` used by `test_import_metapath1` to the
version available in the latest release of `setuptools` (67.8.0).
This version is compatible with PEP 451, and hence python 3.12.

bwoodsend commented:
I'll deal with the extraneous *fizz*.

rokm commented:
Looks like it's also time to implement proper dependency scanning for Tcl/Tk libraries in splash, because the `tcl86t.dll` shipped with Windows build of 3.12b1 depends on `zlib1.dll` and we fail to pick that up...

bwoodsend commented:
This fizz thing looks harmless.

The code it's scanning for calls to functions is:

```python
def foo():
    bar()
    return [fizz(3) for i in range(10)]
```

Until now, the body of a comprehension loop (i.e. the `fizz(3)` bit) was a separate code object nested inside the first but now it has been inlined so that this test fails but as soon as you switch over to recursive scanning (which PyInstaller uses exclusively) you get the same results.

Before:

```python
(Pdb++) foo_code.co_names
('bar', 'range')  # <-- no fizz()
(Pdb++) foo_code.co_consts[1]
<code object <listcomp> at 0x7fd4487b1630, file "<no file>", line 4>  # <-- nested code object
(Pdb++) foo_code.co_consts[1].co_names
('fizz',)  # <-- there's fizz!
```
Verses after:

```python
(Pdb) foo_code.co_names
('bar', 'range', 'fizz')
(Pdb) foo_code.co_consts
(None, 10, 3)
```

bwoodsend committed:
test/bytecode: Update test for scanning nested code objects.

In Python <= 3.11, the compiled bytecode for [foo() for i in range(10)]
is two code objects - the code object for the `foo()` bit being embedded
in the code object for everything else. In Python >= 3.12, it's one flat
code object. This makes it unsuitable as a test for scanning nested code
objects since it's only the correct situation on older Pythons.

Swap out the function call in the body of a comprehension loop with
calling a function in a lambda which still does the right thing with any
Python version.

bwoodsend commented:
Does it still make sense to be dragging distutils support to 3.12 given that the package only sort of exists now?

rokm commented:
> Does it still make sense to be dragging distutils support to 3.12 given that the package only sort of exists now?

I think it does - a simple `import distutils` program should trigger collection of setuptools' version under 3.12, if available.  

But we have whole summer to figure out the best way to handle this (i.e., the one with minimal changes compared to how things work right now and what our current hooks are made to handle)... despite that "high-priority" label

rokm committed:
hooks: add pre-safe-import-module hook for distutils

Starting with python 3.12, `distutils` is not part of stdlib anymore,
and it is provided solely by `setuptools`. Due to limitations of the
modulegraph, the latter cannot find and properly analyze the
`setuptools`-provided `distutils` (and its submodules) under mapped
names, so we need the pre-safe-import-module hook to mark the
`distutils` as a run-time package.

Marking the package as a run-time one means that we also need to
mark its submodules and subpackages as run-time ones.

bwoodsend commented:
Let's merge this now rather than wait for the official 3.12 release? There should be very little difference between Python's beta releases and the real ones.

rokm commented:
I suppose we could, if we're fine with depending on development version of `pywin32-ctypes` for python 3.12. 

Whatever issues may come up between now and October, we would have to deal with either way, and earlier exposure means faster feedback.

bwoodsend committed:
ci: Enable testing with Python 3.12-dev.

bwoodsend reviewed (approved):
Well I approve of the bits you two wrote.

rokm commented:
We still need to update the README and add a news fragment, then it should be ready for merge.

danyeaw committed:
Add Python 3.12 to README, add news fragment

danyeaw commented:
Thanks @bwoodsend and @rokm for the teamwork on this one!
